<div>Hi,</div><div><br></div>OK, no correlation with transcriptomic data yet, but I did do some taxonomic analysis on th metagenome. The raw data is on MG-RAST. If you open an account there, I can share the raw data with you online. <a href="http://metagenomics.nmpdr.org/" target="_blank">http://metagenomics.nmpdr.org/</a><div>

<br></div><div>Basically I took each metagenomic sample and looked for alignments with Greengenes, which is a SSU rRNA database. I then took the top taxa found in all the samples (breastfed and formula-fed). Those were firmicutes, actinobacteria, proteobacteria and bacteroidetes. In two samples there were also a large number of verrucomicrobia: which might be close enough to proteobacteria for us to pool them together, not sure. Will check.</div>

<div><br></div><div>Parameters used:</div><div>sequence identity cutoff to Greengenes: 96%</div><div>alignment length &gt;= 100 bp</div><div><br></div><div>In any case, I pooled all the samples of each treatment together, and looked to see whether there are significant differences between the two populations, using chi-squared test. </div>

<div><br></div><div>Here is the table of the pooled samples for the four top taxa:</div><div><br></div><div><div>                                   FF                  BF</div><div>Firmicutes                   1197                305</div>

<div>Actinobacteria                838               1361</div><div>Proteobacteria                319                659</div><div>Bacteroidetes                   3                  549</div><div><br></div><div>The chi-square for goodness of fit rejected the hypothesis that the samples came from the same population. As you can see, the dominant species in the breast-fed (BF) column are actinobacteria, as opposed to Firmicutes in formula-fed. Also there are hardly any bacteroidetes in the formula-fed infants.</div>

<div><br></div><div>I am attaching two  spreadsheets, one for each sample. THey should be pretty self-explanatory, but do call me. One concern si that a chi-square test for homogeneity within each groups fails: that is, formula-fed babies are not homogeneous in their gut taxa, neither are breastfed. I am not sure that this is a problem though: population variation exists, and we can still &quot;legally&quot; pool the samples. However, I do appreciate a statistician looking at this, and discussing it with me. Let me know if you have any concerns about the way I approached this. </div>

<div><br></div><div>Next: a functional analysis, and comparison with the host transcriptome.</div><div><br></div><div>Cheers,</div><div><br></div><div>Iddo </div><br>-- <br>Iddo Friedberg<br><a href="http://iddo-friedberg.net/contact.html" target="_blank">http://iddo-friedberg.net/contact.html</a><br>


</div>
